youtube_dl/extractor/bilibili.py

```diff
@@ -110,17 +126,17 @@ def _real_extract(self, url):
         mobj = re.match(self._VALID_URL, url)
         video_id = mobj.group('id')
         anime_id = mobj.group('anime_id')
+        page = mobj.group('page') or 1
         webpage = self._download_webpage(url, video_id)
 
         if 'anime/' not in url:
-            cid = self._search_regex(
-                r'cid(?:["\']:|=)(\d+)', webpage, 'cid',
-                default=None
-            ) or compat_parse_qs(self._search_regex(
-                [r'EmbedPlayer\([^)]+,\s*"([^"]+)"\)',
-                 r'EmbedPlayer\([^)]+,\s*\\"([^"]+)\\"\)',
-                 r'<iframe[^>]+src="https://secure\.bilibili\.com/secure,([^"]+)"'],
-                webpage, 'player parameters'))['cid'][0]
+            mobj = re.findall(r'cid(?:["\']:|=)(\d+)', webpage)
```

Comment:
Hi @chongjea ,

I tried this code but it does not work well, maybe Bilibili did some modification?
For example, on your testing page `http://www.bilibili.com/video/av19608864/?p=3`,
the `mobj` here will get
```
[u'7', u'31975229', u'31975229', u'31975230', u'31975231', u'31975232', u'31975233', u'31975231', u'31975231']
```
So you'll get wrong `cid` at next line.

How about doing some change like below?
```python
seen_mobj = set()
mobj = [c for c in mobj if int(c) > 10 and not (c in seen_mobj or seen_mobj.add(c))]
```

Reply:
Try this:
``` mobj = {}
   ll = re.findall(r'cid(?:["\']:|=)(\d+),["\']page(?:["\']:|=)(\d+)', webpage)
   for e in ll:
        mobj[e[1]] = e[0]
   cid =  mobj[page] or \ 
```

Reply:
Is this resolve ?
Can merge to the branch ?

Reply:
Hi @dragonken,

The current commit by @chongjea does not work now.
@atxalex 's code also does not work in my environment.
I am currently using my patch privately.

Should I open a new PR? Seems @chongjea not response anymore...

Reply:
Hi @beards,

The variable name is hard to get the idea.
It looks like I add the following before _download_json work.

`Line 173: video_id = cid`

But the downloaded filename title is not quite right after download.

Reply:
Hi @dragonken ,

Yes, actually I don't really understand the naming from @chongjea , but I know what's the correct data.
What do you mean to set video_id? Is the number correct? Line 173?
`video_id` is not used evermore after line 173.

Reply:
Yes.  The _download_json use video_id.  It seems some other lines are using video_id after that.

My current code as below.
https://hastebin.com/tedofunizu.py

Reply:
Hi @dragonken ,

OK, It's the line number in your code, not in this post.
I see you already adopt my patch, so you should able to download successfully.

And for the `video_id` you mentioned, actually it does not related to content fetching.
You can try to remove it, your code will still work.
The `video_id` is referenced in `InfoExtractor._download_json` that only used for writing message for command line status output and for filename.

I think it's wrong to set it to `cid` because it should be a meaningful ID let user easily refer to a particular video URL, but the `cid` is just a internal ID of a piece of video, that you'll never see it without inspecting the code.

For example, for the url `https://www.bilibili.com/video/av20287268/?p=2`, 
the `video_id` is `20287268` and the `cid` is `33138839`.
You can easily restore the URL from the `video_id`, but you get nothing from the `cid`.

If we have anything can do to improve the `video_id`, I think it should be so:
```python
video_id = 'av{}_p{}'.format(video_id, page)
```